src/kOS/UserIO/TelnetWelcomeMenu.cs

```diff
@@ -25,12 +25,248 @@ public class TelnetWelcomeMenu : MonoBehaviour
         private readonly StringBuilder localMenuBuffer;
         private bool firstTime;
         private volatile bool forceMenuReprint;
-        
+        private MenuLevel currentLevel;
+
+        private abstract class MenuLevel
+        {
+            public virtual MenuLevel parentMenu { get { return new OuterMenu (); } }
+
+            public virtual List<String> helpText { get; } = new List<String> { };
+
+            public virtual List<List<String> > headers { get; } = new List<List<String> > {
+                new List<String> { "Menu", "" },
+                new List<String> { "Pick", "" },
+                new List<String> { "----", "----------------" }
+            };
+            public virtual String format { get; } = "{0,4} {1}";
+
+            public virtual void PrintMenu (TelnetWelcomeMenu parent)
+            {
+
+                List<String> displayChoices = new List<string> { };
+                int longestLength = 0;
+                foreach (List<String> hdrLine in headers) {
+                    String choice = String.Format (format, hdrLine.Cast<object> ().ToArray ());
+                    displayChoices.Add (choice);
+                    longestLength = Math.Max (choice.Length, longestLength);
+                }
+                int i = 1;
+                foreach (object[] bodyLine in this.MenuItems()) {
+                    string choice = String.Format (format, new object[] { "[" + i.ToString () + "]" }.Concat (bodyLine).ToArray ());
+                    displayChoices.Add (choice);
+                    longestLength = Math.Max (choice.Length, longestLength);
+                    i++;
+                }
+                foreach (string choice in displayChoices) {
+                    string choicePaddedToLongest = choice + new String (' ', (longestLength - choice.Length));
+                    parent.telnetServer.Write (parent.CenterPadded (choicePaddedToLongest, ' ') + (char)UnicodeCommand.STARTNEXTLINE);
+                }
+            }
+
+            public virtual List< object[] > MenuItems ()
+            { // Deliberately empty, so this is allowed not to be implemented.
+                return new List< object[] > ();
+            }
+
+            public virtual int ItemCount ()
+            {
+                return MenuItems ().Count;
+            }
+
+            public virtual bool ItemsChanged ()
+            {
+                return false;
+            }
+
+            public virtual void Selected (TelnetWelcomeMenu parent, int k)
+            {
+                return;
+            }
+        }
+
+        private class OuterMenu : MenuLevel
```

Comment:
As to the nested classes, we do actually have one or two places in the mod already where that technique is used.  I'm not as opposed to it as @hvacengi is, provided good justification can be provided for why the nested class is inherently part of the private implementation details of the class its inside of, rather than something the rest of the system could make use of too.

I do find putting two classes in the same file ugly when they're both at the outermost nesting level, though.  Putting them in the same file because they're *nested* is fine because only the outermost class is seen by the rest of the system, and the filename can match that outermost class name.  But when two classes that are both at the outermost level and thus are both seen by the rest of the system, are in the same file, that makes it impossible to have every public class name match the filename the class is inside of.  At most it can match for only one of the two classes.  There were a lot of things about Java that were annoying but one thing I think it definitely got right is that it actually enforced this design principle in the compiler itself.  The directory path layout *IS* the namespace layout with Java.  A class is called foo.bar.baz *because* the compiler found it in path foo/bar/baz.java, not because you added some text saying so.